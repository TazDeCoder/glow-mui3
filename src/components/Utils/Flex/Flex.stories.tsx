import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Flex } from './Flex';
import { Box } from '../Box';

export default {
  title: 'Components/Utils/Flex',
  component: Flex,
} as ComponentMeta<typeof Flex>;

export const Default: ComponentStory<typeof Flex> = (args) => (
  <Flex
    sx={{
      '& > div': {
        flexGrow: 1,
        textAlign: 'center',
        variant: 'headline.sm',
      },
    }}
    width="90%"
    height={240}
    mt={2}
    mx="auto"
    bg="surface.main"
    alignItems="center"
    {...args}
  >
    <Box bg="primary.main" color="primary.onMain">
      Flex
    </Box>
    <Box bg="secondary.main" color="secondary.onMain">
      Box
    </Box>
  </Flex>
);
