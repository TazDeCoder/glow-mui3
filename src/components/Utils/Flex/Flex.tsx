/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef } from 'react';
import { jsx, useTheme, CSSObject, ArrayInterpolation } from '@emotion/react';
import { css, ThemeUICSSProperties } from '@theme-ui/css';

import { Box, BoxOwnProps } from '../Box';
import { pickSystemProps } from '../../shared/utils';
import type { Assign, ForwardRef } from '../../types';

const flexSystemProps = [
  // system props inherited from styled-system (https://styled-system.com/table)
  // flexbox props
  'alignItems',
  'alignContent',
  'justifyItems',
  'justifyContent',
  'flexWrap',
  'flexDirection',
  'flex',
  'flexGrow',
  'flexShrink',
  'flexBasis',
  'justifySelf',
  'alignSelf',
  'order',
  'gap',
  'rowGap',
  'columnGap',
] as const;

type FlexSystemPropsKeys = typeof flexSystemProps[number];
type FlexSystemProps = Pick<ThemeUICSSProperties, FlexSystemPropsKeys>;

export interface FlexOwnProps extends BoxOwnProps, FlexSystemProps {}

export interface FlexProps
  extends Omit<
  Assign<React.ComponentPropsWithRef<'div'>, FlexOwnProps>,
  'ref'
  > {}

const baseStyles: CSSObject = {
  display: 'flex',
};

export const Flex: ForwardRef<any, FlexProps> = forwardRef((props, ref) => {
  const theme = useTheme();

  const systemPropsStyles = css(
    pickSystemProps<FlexOwnProps, FlexSystemPropsKeys>(props, flexSystemProps),
  )(theme);

  const style: ArrayInterpolation<unknown> = [baseStyles, systemPropsStyles];

  const rest: Partial<Omit<FlexProps, FlexSystemPropsKeys>> = {
    ...props,
  };

  for (let i = 0; i < flexSystemProps.length; i += 1) {
    // Remove all system props from props as no longer needed
    flexSystemProps.forEach((name) => {
      delete (rest as Record<string, unknown>)[name];
    });
  }

  return <Box ref={ref} css={style} {...rest} />;
});
