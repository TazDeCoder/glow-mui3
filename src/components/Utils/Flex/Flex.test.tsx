import { render } from '@testing-library/react';

import { Flex } from './Flex';

describe('<Flex />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<Flex>Foo</Flex>);
      expect(container).toMatchSnapshot();
    });
  });
});
