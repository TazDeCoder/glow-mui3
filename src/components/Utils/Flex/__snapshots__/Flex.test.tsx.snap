// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flex /> Snapshots renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="emotion-0"
  >
    Foo
  </div>
</div>
`;
